021-team-credential-managers/proposal.md

```diff
@@ -0,0 +1,128 @@
+  credential has to be duplicated under each team's path.
+
+  By moving credential manager config to each team we can instead leverage the
+  credential manager's access control to determine how credentials are shared
+  across teams (e.g. Vault policies).
+
+* With Vault, this makes it impossible to use any backend except `kv`, because
+  all keys live under the same path scheme, and different backends can't be
+  mounted under paths managed by other backends. This removes a lot of the
+  value of using Vault in the first place.
+
+  By eliminating the path enforcement you can now refer to different secret
+  backend mount points.
+
+* Some credential managers, e.g. Azure KeyVault, have very strict requirements
+  for key names (`[a-z\-]+`), effectively making scoping conventions impossible
+  to enforce.
+
+  By configuring at the team level, each team can point to their own KeyVault
+  or configure their own access control.
+
+* It would be nice to be able to leverage a specialized credential manager like
+  IAM/STS for some things (like the `s3` resource) and use Vault for everything
+  else. Right now you can only configure one credential manager, so this is
+  impossible.
+
+  By allowing teams to configure multiple credential managers, all credential
+  managers can be tried in order when looking up a given credential.
+
+
+# Proposal
+
+The first step is to extend the team config file set by `fly set-team --config`
+to support configuring credential managers. Something like this:
+
+```yaml
+roles: # ...
+
+credential_managers:
+- type: iam
+  config:
+    access_key: blahblah
+    secret_key: blahblah
+- type: vault
+  config:
+    url: https://vault.example.com:8200
+    ca_cert: |
+      -----BEGIN CERTIFICATE-----
+      ...
+    client_token: blahblahbla
+```
+
+Then, any time we're resolving a `((var))` the `web` node would resolve the var
+using each configured credential manager, in order. Distinct fields can be
+accessed like `((foo.bar))`, and nested credential paths can be accessed like
+`((foo/bar/baz))`.
+
+In this case, team's Vault auth config would be associated to a policy which
+determines which credentials the team can access. This way shared credentials
+can be shared without duplicating the credential, and private credentials can
+be kept private.
+
+All credential managers would be modified to remove the automatic team/pipeline
+variable name scoping. They would instead be looked up starting from the root
+level.
+
+
+# Open Questions
+
+* Is there a need for globally-configured and team-configured credential
+  managers to coexist, or can we switch to entirely team-configured (as is the
+  initial goal)?
+
+* Concourse will now be responsible for safely storing access to each and every
+  credential manager, which increases risk. Is it enough to mitigate this by
+  requiring that database encryption be configured?
+
+* Will anyone miss the automatic credential scoping assumptions? Is there value
```

Comment:
Regarding automated scoping:

1. I think it's really important to introduce this as a non-breaking change for existing Concourse users.

    Let's say an existing 4.x or 5.x deployment has 15 teams, 100 pipelines, and the secrets are already arranged as `team/foo` and `team/pipeline/foo` in Vault. Asking end-users to reconfigure all teams after upgrade, and also changing all pipelines to refer to `team/pipeline/foo` instead of `foo`... would be a big stretch

2. Having automated scoping is actually useful.  Right now, when a new pipeline is created, we put quite a few pipeline-specific variables into its own `team/pipeline` "tree" in Vault. That way you can clearly tell which secret belongs to which pipeline, and secrets don't get mixed up in one big pile. Same way, if you have a pipeline called `test`, we don't necessarily want to change all its variables in its template to look like `test/foo` (see previous item too)

3. Also, just to point out, we are sharing a lot of the same secrets between teams. E.g. SMTP host/port/user/pass, etc. With the absence of "global" cred mgr, we'll have to put the same cred mgr config into each team. But we can definitely live with that.

Can scopes be added like this, with `team` and `pipeline` resolved at runtime?

```
credential_managers:
- type: vault
  config:
    url: ...
    ca_cert: ...
    client_token: ...
    prefixes:
      - /concourse
      - /concourse/((team))
      - /concourse/((team))/((pipeline))
```

Reply:
Thanks, those concerns make a lot of sense! I think I'll incorporate this back in some form, then - either as it is today or as something like your example. :+1: 

Conceptually it makes a lot of sense, plus you could always use absolute-path syntax to explicitly refer to a common key.

We might want to avoid a nested path scheme, though - Vault doesn't allow different secret engines to have overlapping mount paths, so the whole `/concourse` path would have to be mounted as `kv`, and then you'd have to use absolute paths to refer to any keys that use a different secret engine. This is the main limitation with today's approach.

Hmm, maybe with your proposal they could just list the same scheme for other engines as other prefixes?:

```
  prefixes:
  - /kv
  - /kv/((team))
  - /kv/((team))/((pipeline))
  - /aws
  - /aws/((team))
  - /aws/((team))/((pipeline))
```

Reply:
just one other thing to point out. when I was refactoring cred mgr code for caching, I noticed that Kubernetes cred mgr also needs a namespace in addition to the path -- it's dynamic and team name is a part of it. Not sure why it was done that way.. but we need to think how it would fit here